slm_lab/agent/net/net_util.py

```diff
@@ -83,34 +85,42 @@ def get_out_dim(body, add_critic=False):
     return out_dim
 
 
-def init_gru_layer(layer):
-    '''Initializes a GRU layer in with xavier_uniform initialization and 0 biases'''
-    for layer_p in layer._all_weights:
-        for p in layer_p:
-            if 'weight' in p:
-                torch.nn.init.xavier_uniform_(layer.__getattr__(p))
-            elif 'bias' in p:
-                torch.nn.init.constant_(layer.__getattr__(p), 0.0)
+def init_layers(net, init_fxn):
+    if init_fxn == 'xavier_uniform_':
+        try:
+            gain = torch.nn.init.calculate_gain(net.hid_layers_activation)
+        except ValueError:
+            gain = 1
+        init_fxn = partial(torch.nn.init.xavier_uniform_, gain=gain)
+    elif 'kaiming' in init_fxn:
+        assert net.hid_layers_activation in ['relu', 'leaky_relu'], f'Kaiming initialization not supported for {net.hid_layers_activation}'
+        init_fxn = torch.nn.init.__dict__[init_fxn]
+        init_fxn = partial(init_fxn, nonlinearity=net.hid_layers_activation)
+    else:
+        init_fxn = torch.nn.init.__dict__[init_fxn]
+    net.apply(partial(init_parameters, init_fxn=init_fxn))
 
 
-def init_layers(layers):
+def init_parameters(module, init_fxn):
     '''
-    Initializes all of the layers of type 'Linear', 'Conv', or 'GRU', using xavier uniform initialization for the weights and 0.01 for the biases, 0.0 for the biases of the GRU.
-    Initializes all layers of type 'BatchNorm' using uniform initialization for the weights and the same as above for the biases
+    Initializes module's weights using init_fxn, which is the name of function from from torch.nn.init
+    Initializes module's biases to either 0.01 or 0.0, depending on module
+    The only exception is BatchNorm layers, for which we use uniform initialization
     '''
     bias_init = 0.01
-    for layer in layers:
-        classname = layer.__class__.__name__
-        if 'BatchNorm' in classname:
-            torch.nn.init.uniform_(layer.weight.data)
-            torch.nn.init.constant_(layer.bias.data, bias_init)
-        elif 'GRU' in classname:
-            init_gru_layer(layer)
-        elif 'Linear' in classname:
-            torch.nn.init.xavier_uniform_(layer.weight.data)
-            torch.nn.init.constant_(layer.bias.data, bias_init)
-        else:
-            pass
+    classname = module.__class__.__name__
+    if 'BatchNorm' in classname:
+        init_fxn(module.weight)
+        torch.nn.init.constant_(module.bias, bias_init)
+    elif 'GRU' in classname:
+        for name, param in module.named_parameters():
+            if 'weight' in name:
+                init_fxn(param)
+            elif 'bias' in name:
+                torch.nn.init.constant_(param, 0.0)
+    elif 'Linear' in classname or ('Conv' in classname and 'Net' not in classname):
+        init_fxn(module.weight)
+        torch.nn.init.constant_(module.bias, bias_init)
```

Comment:
beautiful, this brings proper weight init functionality to the lab!
will merge to staging first, then:
- some minor stylefix (name `fxn` to `fn` per conventions in the lab)
- apply the new init_layers function to all networks